src/PlcServer.cs

```diff
@@ -43,7 +45,209 @@ public PlcServer(OpcPlcConfiguration config, PlcSimulation plcSimulation, TimeSe
+    }
+
+    public override ResponseHeader Publish(
+        RequestHeader requestHeader,
+        SubscriptionAcknowledgementCollection subscriptionAcknowledgements,
+        out uint subscriptionId,
+        out UInt32Collection availableSequenceNumbers,
+        out bool moreNotifications,
+        out NotificationMessage notificationMessage,
+        out StatusCodeCollection results,
+        out DiagnosticInfoCollection diagnosticInfos)
+    {
+        try
+        {
+            OperationContext context = ValidateRequest(requestHeader, RequestType.CreateSubscription);
+
+            var responseHeader = base.Publish(requestHeader, subscriptionAcknowledgements, out subscriptionId, out availableSequenceNumbers, out moreNotifications, out notificationMessage, out results, out diagnosticInfos);
+
+            int events = 0;
+            int dataChanges = 0;
+            int diagnostics = 0;
+            notificationMessage.NotificationData.ForEach(x =>
+            {
+                if (x.Body is DataChangeNotification changeNotification)
+                {
+                    dataChanges += changeNotification.MonitoredItems.Count;
+                    diagnostics += changeNotification.DiagnosticInfos.Count;
+                }
+                else if (x.Body is EventNotificationList eventNotification)
+                {
+                    events += eventNotification.Events.Count;
+                }
+                else
+                {
+                    _logger.LogDebug("Unknown notification type: {notificationType}", x.Body.GetType().Name);
+                }
+            });
+
+            Meters.AddPublishedCount(context.SessionId.ToString(), subscriptionId.ToString(), dataChanges, events);
+
+            _logger.LogDebug("{function} successfully with session: {sessionId} and subscriptionId: {subscriptionId}",
+                nameof(Publish),
+                context.SessionId,
+                subscriptionId);
+
+            return responseHeader;
+        }
+        catch (Exception ex)
+        {
+            Meters.RecordTotalErrors(nameof(Publish), ex.GetType().ToString());
+            _logger.LogError(ex, "Error publishing.");
+            throw;
+        }
+    }
+
+    public override ResponseHeader Read(
+        RequestHeader requestHeader,
+        double maxAge,
+        TimestampsToReturn timestampsToReturn,
+        ReadValueIdCollection nodesToRead,
+        out DataValueCollection results,
+        out DiagnosticInfoCollection diagnosticInfos)
+    {
+        try
+        {
+            var responseHeader = base.Read(requestHeader, maxAge, timestampsToReturn, nodesToRead, out results, out diagnosticInfos);
+
+            _logger.LogDebug("{function} completed successfully", nameof(Read));
+
+            return responseHeader;
+        }
+        catch (Exception ex)
+        {
+            Meters.RecordTotalErrors(nameof(Read), ex.GetType().ToString());
+            _logger.LogError(ex, "Error reading.");

```

Comment:
>. [](http://example.com/codeflow?start=47&length=1)

There are still trailing dots, please remove all of them

Reply:
fixed